sky/provision/aws/instance.py

```diff
@@ -184,20 +174,110 @@ def cleanup_ports(
     }])
     num_sg = len(list(sgs))
     if num_sg == 0:
-        logger.warning(f'Expected security group {sg_name} not found. '
-                       'Skip cleanup.')
-        return
+        logger.warning(f'Expected security group {sg_name} not found. ')
+        return None
     if num_sg > 1:
         # TODO(tian): Better handle this case. Maybe we can check when creating
         # the SG and throw an error if there is already an existing SG with the
         # same name.
-        logger.warning(f'Found {num_sg} security groups with name {sg_name}. '
-                       'Skip cleanup. Please delete them manually.')
+        logger.warning(f'Found {num_sg} security groups with name {sg_name}. ')
+        return None
+    return list(sgs)[0]
+
+
+def _maybe_move_to_new_sg(
+    ec2: Any,
+    cluster_name_on_cloud: str,
+    expected_sg_name: str,
+) -> Any:
+    filters = [
+        {
+            'Name': 'instance-state-name',
+            # exclude 'shutting-down' or 'terminated' states
+            'Values': ['pending', 'running', 'stopping', 'stopped'],
+        },
+        *_cluster_name_filter(cluster_name_on_cloud),
+    ]
+    instances = ec2.instances.filter(Filters=filters)
+    if len(list(instances)) != 1:
+        logger.warning(
+            f'Expected 1 instance with cluster name {cluster_name_on_cloud}, '
+            f'but found {len(list(instances))}. Skip creating security group.')
+        return None
+    instance = list(instances)[0]
+    sg_names = [sg['GroupName'] for sg in instance.security_groups]
+    if len(sg_names) != 1:
+        logger.warning(
+            f'Expected 1 security group for instance {instance.id}, '
+            f'but found {len(sg_names)}. Skip creating security group.')
+        return None
+    sg_name = sg_names[0]
+    if sg_name == expected_sg_name:
+        return _get_sg_from_name(ec2, sg_name)
+    logger.info(f'Cluster {cluster_name_on_cloud} is using shared security '
+                f'group {sg_name}. Change to dedicated security group '
+                f'{expected_sg_name} to allow more inbound rules.')
+    # The security groups will be automatically created by config.py for AWS
+    # since we write it to provider config. But it won't change an existing
+    # instance's security group, so here we move to it.
+    sg = _get_sg_from_name(ec2, expected_sg_name)
+    instance.modify_attribute(Groups=[sg.id])
+    return sg
+
+
+def open_ports(
+    cluster_name_on_cloud: str,
+    ports: List[str],
+    provider_config: Optional[Dict[str, Any]] = None,
+) -> None:
+    """See sky/provision/__init__.py"""
+    assert provider_config is not None, (cluster_name_on_cloud, provider_config)
+    region = provider_config['region']
+    ec2 = _default_ec2_resource(region)
+    sg_name = provider_config['security_group']['GroupName']
```

Comment:
Why don't we use `_filter_instances`?

Reply:
Did you mean the code in L201 (inside `_maybe_move_to_new_sg`)..? I changed that